File: java/sodium/src/sodium/StreamLoop.java
package sodium;

public class StreamLoop<A> extends StreamWithSend<A> {
    boolean assigned = false;

    public StreamLoop()
    {
    	if (Transaction.getCurrentTransaction() == null)
    	    throw new RuntimeException("StreamLoop/CellLoop must be used within an explicit transaction");
    }

    public void loop(Stream<A> ea_out)
    {
        if (assigned)
            throw new RuntimeException("StreamLoop looped more than once");
        assigned = true;
        final StreamLoop<A> me = this;
        unsafeAddCleanup(ea_out.listen_(this.node, new TransactionHandler<A>() {
            public void run(Transaction trans, A a) {
                me.send(trans, a);
            }
        }));
    }
}

File: f#/src/SodiumFRP.FSharp/_Root/StreamLoop.fs
namespace SodiumFRP.FSharp

type StreamLoop<'T> internal () =
    inherit Stream<'T>()
    
    let isAssignedLock = obj ()
    let mutable isAssigned = false
    
    member internal this.Loop transaction (stream : Stream<'T>) =
        lock isAssignedLock (fun () ->
            if isAssigned then invalidOp "Loop was looped more than once."
            isAssigned <- true)
        this.AttachListener <| stream.Listen this.Node this.Send |> ignore
        lock stream.KeepListenersAlive (fun () -> stream.KeepListenersAlive.Use this.KeepListenersAlive)
